adr/adr-002.md

```diff
@@ -0,0 +1,342 @@
+
+// DecreaseTotalLiquidStakedTokens decrements the total liquid staked tokens
+// if the global cap is enabled
+func (k Keeper) DecreaseTotalLiquidStakedTokens(ctx sdk.Context, amount sdk.Int) 
+
+// SafelyIncreaseValidatorTotalLiquidShares increments the total liquid shares on a validator
+// if the validator bond factor is enabled and is not surpassed by this delegation
+func (k Keeper) SafelyIncreaseValidatorTotalLiquidShares(ctx sdk.Context, validator types.Validator, shares sdk.Dec) error 
+
+// DecreaseValidatorTotalLiquidShares decrements the total liquid shares on a validator
+// if the validator bond factor is enabled
+func (k Keeper) DecreaseValidatorTotalLiquidShares(ctx sdk.Context, validator types.Validator, shares sdk.Dec) 
+
+// SafelyDecreaseValidatorBond decrements the total validator's self bond
+// so long as it will not cause the current delegations to exceed the threshold
+// set by validator bond factor
+func (k Keeper) SafelyDecreaseValidatorBond(ctx sdk.Context, validator types.Validator, shares sdk.Dec) error 
+```
+
+### Accounting
+Tracking the total liquid stake and total liquid validator shares requires additional accounting changes in the following transactions/events:
+
+```go
+func Delegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Increment total liquid staked
+    //    Increment validator liquid shares
+}
+
+func Undelegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Decrement total liquid staked
+    //    Decrement validator liquid shares
+}
+
+func BeginRedelegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Decrement source validator liquid shares
+    //    Increment destination validator liquid shares
+}
+
+func TokenizeShares() {
+    ...
+    // If delegator is a NOT liquid staking provider (otherwise the shares are already included)
+    //    Increment total liquid staked
+    //    Increment validator liquid shares
+}
+
+func RedeemTokens() {
+    ...
+    // If delegator is a NOT liquid staking provider 
+    //    Decrement total liquid staked
+    //    Decrement validator liquid shares
+}
+
+func Slash() {
+    ...
+    // Decrement total liquid staked (since total liquid stake is denominated in tokens, not shares)
+}
+```
+
+### Transaction failure cases
+With the liquid staking caps in consideration, there are additional scenarios that should cause a transaction to fail:
+```go
+
+func Delegate() {
+    ...
+    // If delegator is a liquid staking provider
+    //    Fail transaction if delegation exceeds global liquid staking cap
+    //    Fail transaction if delegation exceeds validator bond cap
+}
+
+func Undelegate() {
+    ...
+    // If the unbonded delegation is a ValidatorBond
+    //    Fail transaction if the reduction in validator bond would cause the
+    //    existing liquid delegation to exceed the cap
```

Comment:
I think this might cause normal delegator not able to withdraw.
E.g. liquid staking global cap 30%, total delegation 100K tokens and 30K from liquid staking, 70K from normal staking.
A normal delegator who delegated 10K tokens would like to undelegate.
The delegation vs global cap changes from 30%(30K/100K) -> 33% (30K/90K) - normal users would not be able to withdraw as well if liquid staking cap riches.

Reply:
This is specifically for validator bonds (i.e. delegations that have been labeled as a validator bond), so it should have no impact on normal users unbondings